specs/PushNotifications/PushNotifications-spec.md

```diff
@@ -0,0 +1,400 @@
+* The app will query ApplifeCycle APIs to retrieve an ActivationKind. (Note: This is covered in a seperate AppLifeCycle API spec.)
+* If the Activation is Push, the app will QI the Arguments property to retrieve an instance of PushActivatedEventArgs and get the Push payload from it. Care needs to be taken to Take a Deferral and Release the Deferral while processing the payload for Power Management compliance.
+* If the Activation is Foreground, the app will do 2 things:
+  * It will subscribe to a Foreground In-memory Push event
+  * It will request a Push Channel Asynchronously with an implementation of Progress and Completed event handler.
+    * Expect Channel operations to take around 2-16 minutes in some rare cases (retryable errors). In the 90th percentile case, it should be fairly quick operation (in a few seconds).  
+
+```cpp
+int main()
+{
+    PushActivatorDetails details(winrt::guid("BACCFA91-F1DE-4CA2-B80E-90BE66934EC6"));
+
+    // Registers a Push Trigger and sets up an inproc COM Server for Activations
+    PushManager::RegisterActivator(details, PushRegistrationType::PushTrigger | PushRegistrationType::ComActivator);
+
+    // Check to see if the WinMain activation is due to a Push Activator
+    // Note: This API is currently not present in Reunion but will be included as part of the AppLifecycle Public API spec.
+    auto activation = AppLifecycle::Activation();
+    auto kind = activation.Kind();
+
+    if (kind == ReunionActivationKind::Push)
+    {
+        auto pushArgs = activation.Arguments().try_as<PushReceivedEventArgs>();
+
+        // Call TakeDeferral to ensure that code runs in low power
+        pushArgs.GetDeferral();
+
+        auto payload = pushArgs.Payload();
+
+        // Do stuff to process the raw payload
+
+        // Call CompleteDeferral as good practise: Needed mainly for low power usage
+        pushArgs.CompleteDeferral();
+    }
+    else if (kind == ReunionActivationKind::Launch) // This indicates that the app is launching in the foreground
+    {
+        // Register an event to Intercept Push payloads
+        auto eventToken = details.PushReceived([](const auto&, PushReceivedEventArgs args)
+        {
+                // Call TakeDeferral to ensure that code runs in low power
+                args.GetDeferral();
+
+                auto payload = args.Payload();
+
+                // Do stuff to process the raw payload
+
+                // Call CompleteDeferral as good practise: Needed mainly for low power usage
+                args.CompleteDeferral();
+        });
+
+        auto channelOperation = PushManager::CreateChannelAsync(winrt::guid("F80E541E-3606-48FB-935E-118A3C5F41F4"));
+
+        // Setup the inprogress event handler
+        channelOperation.Progress(
+            [](
+                IAsyncOperationWithProgress<ChannelResult, ChannelResult> const& /* sender */,
+                ChannelResult const& args)
```

Comment:
I recall that there was some rule that progress must be a struct, not a class. Elliot may remember why we had this rule. Maybe the rule isn't relevant any more.

Reply:
@oldnewthing @jonwis Unfortunately, if I change ChannelResult to a struct from a runtimeclass, the compiler complains because structs are not supposed to contain pointer types. In this case, the pointer type is of type Windows.Networking.PushNotifications.PushNotificationChannel. Please advise.

Reply:
@ShawnSteele for guidance as well

Reply:
One possibility is to simply return a combo of the ChannelStatus and ExtendedError code in a seperate PushNotificationChannelProgress struct.

Reply:
The way progress works, you hook up the `Progress` delegate and are handed the original operation (the sender) and the progress information.  For linear progress (like copying a file) a `UInt64` is sufficient - the number of bytes copied so far.

We're suggesting that you change the type of the async operation to be like this (please use the agreed names; I'm just using shorthand here.)

```c#
runtimeclass ChannelManager {
    // The result of the operation is the CreateChannelResult; the progress
    // information is the Status, which moves from "started" through "completed"
    IAsyncOperationWithProgress<CreateChannelResult, CreateChannelStatus> CreateChannelAsync(...);
}
```

So your progress handler looks like:

```c++
auto op = ChannelManager::CreateChannelAsync(...);
op.Progress([](IAsyncOperationWithProgress<CreateChannelResult, CreateChannelStatus> const& op, CreateChannelStatus status)
{
    // Check the 'status' parameter for how far through the process you are
    // Async ops with progress are supposed to support calling op.GetResult() to retrieve
    // an intermediate state result, but I'm not sure how C++/WinRT supports that
});
```

Hey @kennykerr - per https://osgwiki.com/wiki/WinRT_API_Design_Guidelines#Getting_The_Results_of_an_Asynchronous_Operation (sorry non-Microsoft people!) calls to `IAsyncOperationWithProgress<T,V>::GetResults()` are supposed to return any progress generated so far. How does that work with a coroutine-based C++/WinRT implementation?

Reply:
Of course it supports it - https://docs.microsoft.com/en-us/windows/uwp/cpp-and-winrt-apis/concurrency-2#reporting-progress ... super neat!

Reply:
Here's an example: https://docs.microsoft.com/en-us/windows/uwp/cpp-and-winrt-apis/concurrency-2

I will just add that the `-WithProgress` variants are overrated. They are a little clumsy and hard to use in most languages, and frankly seldom used for that reason. As such, I would recommend sticking with either `IAsyncAction` or `IAsyncOperation<T>`. If you must use them I would just note that there is nothing that says they are required to actually provide progress notifications nor that the caller must listen for them. So even if you're stuck implementing or consuming them, you can simply ignore the progress machinery. If you prefer to implement them then [this tells you how](https://docs.microsoft.com/en-us/windows/uwp/cpp-and-winrt-apis/concurrency-2).

Reply:
Understood. I'll take a look at the doc and make relevant updates! Thanks for the tip

Reply:
@jonwis C++/WinRT supports reporting progress, but [you cannot call `GetResult()` until the operation has completed](https://github.com/microsoft/cppwinrt/blob/fc9b4d780f389dec229bb9efe6f04b8aaa099273/strings/base_coroutine_foundation.h#L457):

```cpp
        auto GetResults()
        {
            slim_lock_guard const guard(m_lock);

>>>>>>>>>>>>if (m_status.load(std::memory_order_relaxed) == AsyncStatus::Completed)
            {
                return static_cast<Derived*>(this)->get_return_value();
            }

            rethrow_if_failed();
            WINRT_ASSERT(m_status.load(std::memory_order_relaxed) == AsyncStatus::Started);
>>>>>>>>>>>>throw hresult_illegal_method_call();
        }
```

I'll add "partial GetResults() support" to my C++/WinRT backlog. I don't think it'll be too hard. Would go something like

```cpp
IAsyncOperationWithProgress<double, double> CalcPiTo5DPs()
{
    auto progress{ co_await winrt::get_progress_token() };

    co_await 1s;
    double pi_so_far{ 3.1 };
    progress.partial_result(pi_so_far); // new
    progress(0.2);

    co_await 1s;
    pi_so_far += 4.e-2;
    progress.partial_result(pi_so_far); // new
    progress(0.4);

    co_await 1s;
    pi_so_far += 1.e-3;
    progress.partial_result(pi_so_far); // new
    progress(0.6);

    co_await 1s;
    pi_so_far += 5.e-4;
    progress.partial_result(pi_so_far); // new
    progress(0.8);

    co_await 1s;
    pi_so_far += 9.e-5;
    progress.partial_result(pi_so_far); // new    progress(1.0);
    co_return pi_so_far;
}
```

Reply:
Partial results are now in C++/WinRT. microsoft/cppwinrt#870

Reply:
@oldnewthing Do you happen to know which nuget version of c++/winrt would include this functionality? 

Also from your PR, it sounds like the recommendation is to call partial_return for inprogress updates in addition to the progress calls? And the consumer simply calls into GetResult to retrieve the partial result?
@sharath2727 FYI

Reply:
@kennykerr would be great to push a release with this change so that we can consume it in our code

Reply:
2.0.210309.3 has been published.